### PR TITLE
Creating namespace before secret creation - fixing PR#2156

### PR DESCRIPTION
**Pull request by jaroslaw-pieszka**
**Description**
Implementation in PR#2156 assumed the namespace is already created. Assumption was wrong.

Changes proposed in this pull request:

- Creation of `kyma-system` namespace before secret creation

**Related issue(s)**

### CODE REVIEW

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go [hunk 1]
```diff
@@ -129,14 +129,20 @@ func updateSecretData(secret *apicorev1.Secret, secretFromParameters *apicorev1.
 
 func (s *InjectBTPOperatorCredentialsStep) createOrRetry(k8sClient client.Client, newSecret *apicorev1.Secret, err error, log logrus.FieldLogger) error {
 	if apierrors.IsNotFound(err) {
+		namespace := &apicorev1.Namespace{ObjectMeta: v1.ObjectMeta{Name: secretNamespace}}
+		err = k8sClient.Create(context.Background(), namespace)
+		if err != nil {
```
Comment (szwedm) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:134]: namespace may already exist, maybe it's worth to handle such a case differently to skip the warning in logs.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go [hunk 2]
```diff
@@ -129,14 +129,20 @@ func updateSecretData(secret *apicorev1.Secret, secretFromParameters *apicorev1.
...
+			log.Warnf("could not create %s namespace: %w", secretNamespace, err)
+		}
+
 		err = k8sClient.Create(context.Background(), newSecret)
```
Comment (szwedm) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:138]: how about returning an error earlier if namespace creation fails and it doesn't exist in the cluster?

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step_test.go [hunk 3]
```diff
@@ -65,6 +66,10 @@ func TestInjectBTPOperatorCredentialsStep(t *testing.T) {
 		operation := fixture.FixProvisioningOperation("operation-id", "inst-id")
 		operation.RuntimeID = ""
 
+		namespaces := apicorev1.NamespaceList{}
```
Comment (szwedm) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step_test.go:69]: why do we need the namespaces here?

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go (hunk 2 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:138]: We will fail on secret creation eventually with a message specifying the reason (missing namespace). I have deliberately simplified the logic.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step_test.go (hunk 3 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step_test.go:69]: I think this is leftover to be removed.

Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step_test.go:69]: Removed in the next commit.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go (hunk 1 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:134]: In the next push.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go (hunk 2 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:142]: Will do.

Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:143]: Done.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go [hunk 4]
```diff
@@ -129,14 +129,25 @@ func updateSecretData(secret *apicorev1.Secret, secretFromParameters *apicorev1.
...
+			}
+		}
+
 		err = k8sClient.Create(context.Background(), newSecret)
 		if err == nil {
 			log.Info("the secret for BTP Manager created")
 			return nil
 		}
-		log.Error("failed to create the secret for BTP Manager")
+		log.Errorf("failed to create the secret for BTP Manager: %w", err)
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:148]: when logging wrapping is not needed using %w, just use %s

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go [hunk 5]
```diff
@@ -110,7 +110,7 @@ func (s *InjectBTPOperatorCredentialsStep) createOrUpdateSecret(k8sClient client
 	updateSecretData(&clusterSecret, parametersBasedSecret)
 	err = k8sClient.Update(context.Background(), &clusterSecret)
 	if err != nil {
-		log.Error("failed to update the secret for BTP Manager")
+		log.Errorf("failed to update the secret for BTP Manager: %w", err)
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:113]: when logging wrapping is not needed using %w, just use %s

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go [hunk 6]
```diff
@@ -129,14 +129,25 @@ func updateSecretData(secret *apicorev1.Secret, secretFromParameters *apicorev1.
 
 func (s *InjectBTPOperatorCredentialsStep) createOrRetry(k8sClient client.Client, newSecret *apicorev1.Secret, err error, log logrus.FieldLogger) error {
 	if apierrors.IsNotFound(err) {
+		namespace := &apicorev1.Namespace{ObjectMeta: v1.ObjectMeta{Name: secretNamespace}}
+		err = k8sClient.Create(context.Background(), namespace)
+		if err != nil {
+			if apierrors.IsAlreadyExists(err) {
```
Comment (szwedm) [inline components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:135]: I would check if err is different from AlreadyExists error then use log to provide extra info with details why Create() method failed. When the error is AlreadyExists error, skip the log as the required namespace exists (we don't care who/what created it) and try to create the Secret in existing namespace.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go (hunk 5 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:113]: In the next commit.

File: components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go (hunk 6 above)
Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go:135]: Considering we will use the same step for update/upgrade you are right.

Comment (jaroslaw-pieszka) [reply components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go]: changed

**Comment by jaroslaw-pieszka** [comment]
Test for own_cluster plan
```
k get secret -n kyma-system sap-btp-manager -o json
{
    "apiVersion": "v1",
    "data": {
        "clientid": "dGVzdENsaWVudElE",
        "clientsecret": "dGVzdENsaWVudFNlY3JldA==",
        "cluster_id": "Yjg3MjdhY2ItZDI2MS00ZTMyLWExOTctMjA5OWZkNTNmNjUy",
        "sm_url": "aHR0cHM6Ly9zZXJ2aWNlLW1hbmFnZXIua3ltYS5jb20=",
        "tokenurl": "aHR0cHM6Ly90ZXN0LmF1dGguY29t"
    },
    "kind": "Secret",
    "metadata": {
        "annotations": {
            "Warning": "This secret is generated. Do not edit!"
        },
        "creationTimestamp": "2022-10-29T07:39:09Z",
        "labels": {
            "app.kubernetes.io/managed-by": "kcp-kyma-environment-broker"
        },
        "name": "sap-btp-manager",
        "namespace": "kyma-system",
        "resourceVersion": "202669",
        "uid": "9720354d-8ac2-48fe-af7a-8959161954e3"
    },
    "type": "Opaque"
}
I567567@D659KT66YK update-secret % echo "dGVzdENsaWVudElE"|base64 -D
testClientID% ```

**Comment by jaroslaw-pieszka** [comment]
```
 k get secrets sap-btp-manager -n kyma-system
NAME              TYPE     DATA   AGE
sap-btp-manager   Opaque   5      33m
I567567@D659KT66YK update-secret % k get secrets sap-btp-manager -n kyma-system -o json
{
    "apiVersion": "v1",
    "data": {
        "clientid": "dGVzdENsaWVudElE",
        "clientsecret": "dGVzdENsaWVudFNlY3JldA==",
        "cluster_id": "ZmQ1MmUyY2QtMGQyMi00MjU2LThmZjYtOGQ0MTMyY2UxNTM5",
        "sm_url": "aHR0cHM6Ly9zZXJ2aWNlLW1hbmFnZXIua3ltYS5jb20=",
        "tokenurl": "aHR0cHM6Ly90ZXN0LmF1dGguY29t"
    },
    "kind": "Secret",
    "metadata": {
        "annotations": {
            "Warning": "This secret is generated. Do not edit!"
        },
        "creationTimestamp": "2022-10-29T08:24:39Z",
        "labels": {
            "app.kubernetes.io/managed-by": "kcp-kyma-environment-broker"
        },
        "name": "sap-btp-manager",
        "namespace": "kyma-system",
        "resourceVersion": "3377",
        "uid": "ef8653ee-c76c-4b41-8dee-06233a219351"
    },
    "type": "Opaque"
}
```

**Review by szwedm — APPROVED** [review]
/lgtm

**Pull request merged**